File: crates/txline/src/solana/pda.rs
//! Devnet PDA helpers.

use solana_sdk::pubkey::Pubkey;

use crate::config::{DEVNET_PROGRAM_ID, DEVNET_TXL_MINT, DEVNET_USDT_MINT};
use crate::{Result, TxlineError};

pub const TOKEN_2022_PROGRAM_ID: &str = "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb";
pub const LEGACY_TOKEN_PROGRAM_ID: &str = "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA";
pub const ASSOCIATED_TOKEN_PROGRAM_ID: &str = "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL";
pub const SYSTEM_PROGRAM_ID: &str = "11111111111111111111111111111111";
pub const COMPUTE_BUDGET_PROGRAM_ID: &str = "ComputeBudget111111111111111111111111111111";

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Pda {
    pub address: Pubkey,
    pub bump: u8,
}

#[derive(Debug, Clone, Copy)]
pub struct DevnetPdas {
    pub program_id: Pubkey,
    pub txl_mint: Pubkey,
    pub usdt_mint: Pubkey,
}

impl DevnetPdas {
    pub fn new() -> Result<Self> {
        Ok(Self {
            program_id: parse_pubkey(DEVNET_PROGRAM_ID)?,
            txl_mint: parse_pubkey(DEVNET_TXL_MINT)?,
            usdt_mint: parse_pubkey(DEVNET_USDT_MINT)?,
        })
    }

    pub fn pricing_matrix(&self) -> Pda {
        find_pda(&[b"pricing_matrix"], &self.program_id)
    }

    pub fn token_treasury_v2(&self) -> Pda {
        find_pda(&[b"token_treasury_v2"], &self.program_id)
    }

    pub fn usdt_treasury(&self) -> Pda {
        find_pda(&[b"usdt_treasury"], &self.program_id)
    }

    pub fn token_treasury_vault_ata(&self) -> Result<Pda> {
        token_2022_associated_token_address(&self.token_treasury_v2().address, &self.txl_mint)
    }

    pub fn usdt_treasury_vault_ata(&self) -> Result<Pda> {
        token_2022_associated_token_address(&self.usdt_treasury().address, &self.usdt_mint)
    }

    pub fn user_txl_ata(&self, user: &Pubkey) -> Result<Pda> {
        token_2022_associated_token_address(user, &self.txl_mint)
    }

    pub fn user_usdt_ata(&self, user: &Pubkey) -> Result<Pda> {
        token_2022_associated_token_address(user, &self.usdt_mint)
    }

    pub fn daily_scores_roots(&self, epoch_day: u16) -> Pda {
        let day = epoch_day.to_le_bytes();
        find_pda(&[b"daily_scores_roots", &day], &self.program_id)
    }

    pub fn daily_batch_roots(&self, epoch_day: u16) -> Pda {
        let day = epoch_day.to_le_bytes();
        find_pda(&[b"daily_batch_roots", &day], &self.program_id)
    }

    /// PDA used by the Devnet IDL account `daily_odds_merkle_roots`.
    ///
    /// Upstream docs derive odds validation roots with the `daily_batch_roots`
    /// seed, even though the validation instruction account is named
    /// `daily_odds_merkle_roots`.
    pub fn daily_odds_merkle_roots(&self, epoch_day: u16) -> Pda {
        self.daily_batch_roots(epoch_day)
    }

    pub fn ten_daily_fixtures_roots(&self, epoch_day: u16) -> Pda {
        let aligned = epoch_day - (epoch_day % 10);
        let day = aligned.to_le_bytes();
        find_pda(&[b"ten_daily_fixtures_roots", &day], &self.program_id)
    }
}

pub fn token_2022_associated_token_address(owner: &Pubkey, mint: &Pubkey) -> Result<Pda> {
    let token_program = parse_pubkey(TOKEN_2022_PROGRAM_ID)?;
    let associated_program = parse_pubkey(ASSOCIATED_TOKEN_PROGRAM_ID)?;
    Ok(find_pda(
        &[owner.as_ref(), token_program.as_ref(), mint.as_ref()],
        &associated_program,
    ))
}

pub fn parse_pubkey(value: &str) -> Result<Pubkey> {
    value
        .parse::<Pubkey>()
        .map_err(|err| TxlineError::solana(format!("invalid pubkey {value}: {err}")))
}

fn find_pda(seeds: &[&[u8]], program_id: &Pubkey) -> Pda {
    let (address, bump) = Pubkey::find_program_address(seeds, program_id);
    Pda { address, bump }
}

File: typescript/src/solana/pda.ts
import { getProgramDerivedAddress } from "@solana/kit";
import {
  DEVNET_PROGRAM_ID,
  DEVNET_TXL_MINT,
  DEVNET_USDT_MINT,
} from "../config.js";
import {
  addressBytes,
  toAddress,
  type AddressLike,
} from "./types.js";
import type { Address } from "@solana/kit";

export const TOKEN_2022_PROGRAM_ID =
  "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb";
export const LEGACY_TOKEN_PROGRAM_ID =
  "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA";
export const ASSOCIATED_TOKEN_PROGRAM_ID =
  "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL";
export const SYSTEM_PROGRAM_ID = "11111111111111111111111111111111";
export const COMPUTE_BUDGET_PROGRAM_ID =
  "ComputeBudget111111111111111111111111111111";

export interface Pda {
  readonly address: Address;
  readonly bump: number;
}

export class DevnetPdas {
  readonly programId = toAddress(DEVNET_PROGRAM_ID);
  readonly txlMint = toAddress(DEVNET_TXL_MINT);
  readonly usdtMint = toAddress(DEVNET_USDT_MINT);

  async pricingMatrix(): Promise<Pda> {
    return await findPda(["pricing_matrix"], this.programId);
  }

  async tokenTreasuryV2(): Promise<Pda> {
    return await findPda(["token_treasury_v2"], this.programId);
  }

  async usdtTreasury(): Promise<Pda> {
    return await findPda(["usdt_treasury"], this.programId);
  }

  async tokenTreasuryVaultAta(): Promise<Pda> {
    return await token2022AssociatedTokenAddress(
      (await this.tokenTreasuryV2()).address,
      this.txlMint,
    );
  }

  async usdtTreasuryVaultAta(): Promise<Pda> {
    return await token2022AssociatedTokenAddress(
      (await this.usdtTreasury()).address,
      this.usdtMint,
    );
  }

  async userTxlAta(user: AddressLike): Promise<Pda> {
    return await token2022AssociatedTokenAddress(user, this.txlMint);
  }

  async userUsdtAta(user: AddressLike): Promise<Pda> {
    return await token2022AssociatedTokenAddress(user, this.usdtMint);
  }

  async dailyScoresRoots(epochDay: number): Promise<Pda> {
    return await findPda(
      ["daily_scores_roots", u16Le(epochDay)],
      this.programId,
    );
  }

  async dailyBatchRoots(epochDay: number): Promise<Pda> {
    return await findPda(
      ["daily_batch_roots", u16Le(epochDay)],
      this.programId,
    );
  }

  async dailyOddsMerkleRoots(epochDay: number): Promise<Pda> {
    return await this.dailyBatchRoots(epochDay);
  }

  async tenDailyFixturesRoots(epochDay: number): Promise<Pda> {
    const aligned = epochDay - (epochDay % 10);
    return await findPda(
      ["ten_daily_fixtures_roots", u16Le(aligned)],
      this.programId,
    );
  }
}

export async function findPda(
  seeds: readonly (string | Uint8Array)[],
  programAddress: AddressLike,
): Promise<Pda> {
  const [pda, bump] = await getProgramDerivedAddress({
    programAddress: toAddress(programAddress),
    seeds: seeds.map((seed) =>
      typeof seed === "string" ? new TextEncoder().encode(seed) : seed,
    ),
  });
  return { address: pda, bump };
}

export async function token2022AssociatedTokenAddress(
  owner: AddressLike,
  mint: AddressLike,
): Promise<Pda> {
  return await findPda(
    [
      addressBytes(owner),
      addressBytes(TOKEN_2022_PROGRAM_ID),
      addressBytes(mint),
    ],
    ASSOCIATED_TOKEN_PROGRAM_ID,
  );
}

export function u16Le(value: number): Uint8Array {
  if (!Number.isInteger(value) || value < 0 || value > 0xffff) {
    throw new RangeError("epoch day must fit in u16");
  }
  return Uint8Array.of(value & 0xff, (value >> 8) & 0xff);
}
